packages/next/src/build/validate-app-paths.ts

```diff
@@ -0,0 +1,281 @@
+import {
+  getParamProperties,
+  type SegmentParam,
+} from '../shared/lib/router/utils/get-segment-param'
+import {
+  isInterceptionAppRoute,
+  parseAppRoute,
+  type NormalizedAppRoute,
+  type NormalizedAppRouteSegment,
+} from '../shared/lib/router/routes/app'
+
+/**
+ * Validates segment parameters for common syntax errors.
+ * Based on validation logic from sorted-routes.ts
+ */
+function validateSegmentParam(param: SegmentParam, pathname: string): void {
+  // Check for empty parameter names
+  if (param.param.length === 0) {
+    throw new Error(`Parameter names cannot be empty in route "${pathname}".`)
+  }
+
+  // Check for three-dot character (…) instead of ...
+  if (param.param.includes('…')) {
+    throw new Error(
+      `Detected a three-dot character ('…') in parameter "${param.param}" in route "${pathname}". Did you mean ('...')?`
+    )
+  }
+
+  // Check for optional non-catch-all segments (not yet supported)
+  if (
+    param.type !== 'optional-catchall' &&
+    param.param.startsWith('[') &&
+    param.param.endsWith(']')
+  ) {
+    throw new Error(
+      `Optional route parameters are not yet supported ("[${param.param}]") in route "${pathname}".`
+    )
+  }
+
+  // Check for extra brackets
+  if (param.param.startsWith('[') || param.param.endsWith(']')) {
+    throw new Error(
+      `Segment names may not start or end with extra brackets ('${param.param}') in route "${pathname}".`
+    )
+  }
```

Comment:
Is this duplication of runtime validation? Seems like we might be duplicating a lot and we might benefit from using the same validation in dev as in prod

Reply:
The only place that does this today is the route sorting right now. I'd like to reuse this in dev, was planning a follow up for it.